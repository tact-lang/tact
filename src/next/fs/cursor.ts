import type { RelativePath } from "@/next/fs/path";

/**
 * Points to a node of a file system: directory or file
 */
export interface Cursor {
    /**
     * Move cursor by given relative path
     */
    readonly focus: (path: RelativePath) => this;

    /**
     * Read file under the cursor
     */
    readonly read: () => Promise<string | undefined>;

    /**
     * Write to file under the cursor
     */
    readonly write: (content: string | Blob) => Promise<void>;

    /**
     * Get full absolute path to a file or directory
     *
     * NB! Only to be used with `log.source()`
     */
    readonly getAbsolutePathForLog: () => string;
}
