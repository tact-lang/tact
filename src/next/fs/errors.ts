import type { Logger } from "@/error/logger-util";

export const FsErrors = <M, R>(l: Logger<M, R>) => {
    const regular = {
        ENOENT: (absPath: string) => {
            return l.error(l.text`File not found: ${l.path(absPath)}`);
        },
        EACCES: (absPath: string) => {
            return l.error(l.text`Permission denied: ${l.path(absPath)}`);
        },
        EISDIR: (absPath: string) => {
            return l.error(l.text`Directory is not a file: ${l.path(absPath)}`);
        },
        EMFILE: (absPath: string) => {
            return l.error(l.text`Too many open files: ${l.path(absPath)}`);
        },
        ENAMETOOLONG: (absPath: string) => {
            return l.error(l.text`Filename too long: ${l.path(absPath)}`);
        },
        ENOSPC: (absPath: string) => {
            return l.error(l.text`No space left on device: ${l.path(absPath)}`);
        },
    } as const;
    return {
        regular,
        unexpected: (error: unknown) => {
            throw error;
        },
        outOfRoot: (absPath: string) => {
            return l.error(
                l.text`Outside of the root directory: ${l.path(absPath)}`,
            );
        },
        readonly: (absPath: string) => {
            return l.error(l.text`Filesystem is readonly: ${l.path(absPath)}`);
        },
    };
};

export type FsErrors = ReturnType<typeof FsErrors>;
