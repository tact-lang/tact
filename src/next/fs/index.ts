export * from "@/next/fs/cursor";
export * from "@/next/fs/path";
export * from "@/next/fs/util";
export * from "@/next/fs/memory-fs";
export * from "@/next/fs/proxy-fs";
