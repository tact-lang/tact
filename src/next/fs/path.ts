/**
 * Safe relative path
 */
export type RelativePath = {
    /**
     * Number of "../" in front of path
     */
    readonly stepsUp: number;
    /**
     * /-separated strings that go after optional ../
     */
    readonly segments: readonly string[];
};
