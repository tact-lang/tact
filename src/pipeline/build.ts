import { beginCell, Cell, Dictionary } from '@ton/core';
import { decompileAll } from '@tact-lang/opcode';
import { writeTypescript } from '../bindings/writeTypescript';
import { featureEnable } from '../config/features';
import { ConfigProject } from "../config/parseConfig";
import { CompilerContext } from "../context";
import { funcCompile } from '../func/funcCompile';
import { writeReport } from '../generator/writeReport';
import { getRawAST } from '../grammar/store';
import files from '../imports/stdlib';
import { consoleLogger, TactLogger } from '../logger';
import { PackageFileFormat } from '../packaging/fileFormat';
import { packageCode } from '../packaging/packageCode';
import { createABITypeRefFromTypeRef } from '../types/resolveABITypeRef';
import { getContracts, getType } from '../types/resolveDescriptors';
import { errorToString } from '../utils/errorToString';
import { posixNormalize } from '../utils/filePath';
import { createVirtualFileSystem } from '../vfs/createVirtualFileSystem';
import { VirtualFileSystem } from '../vfs/VirtualFileSystem';
import { compile } from './compile';
import { precompile } from "./precompile";
import { getCompilerVersion } from './version';

export async function build(args: {
    config: ConfigProject,
    project: VirtualFileSystem,
    stdlib: string | VirtualFileSystem,
    logger?: TactLogger | null | undefined
}) {

    const { config, project } = args;
    const stdlib = (typeof args.stdlib === 'string') ? createVirtualFileSystem(args.stdlib, files) : args.stdlib;
    const logger: TactLogger = args.logger || consoleLogger;

    // Configure context
    let ctx: CompilerContext = new CompilerContext({ shared: {} });
    const cfg: string = JSON.stringify({
        entrypoint: config.path,
        options: (config.options || {})
    });
    if (config.options) {
        if (config.options.debug) {
            logger.error('   > ðŸ‘€ Enabling debug');
            ctx = featureEnable(ctx, 'debug');
        }
        if (config.options.masterchain) {
            logger.error('   > ðŸ‘€ Enabling masterchain');
            ctx = featureEnable(ctx, 'masterchain');
        }
        if (config.options.external) {
            logger.error('   > ðŸ‘€ Enabling external');
            ctx = featureEnable(ctx, 'external');
        }
        if (config.options.experimental && config.options.experimental.inline) {
            logger.error('   > ðŸ‘€ Enabling inline');
            ctx = featureEnable(ctx, 'inline');
        }
    }

    // Precompile
    try {
        ctx = precompile(ctx, project, stdlib, config.path);
    } catch (e) {
        logger.error('Tact compilation failed');
        logger.error(errorToString(e));
        return false;
    }

    // Compile contracts
    let ok = true;
    const built: {
        [key: string]: {
            codeBoc: Buffer,
            // codeFunc: string,
            // codeFift: string,
            // codeFiftDecompiled: string,
            abi: string
        }
    } = {};
    for (const contract of getContracts(ctx)) {
        const pathAbi = project.resolve(config.output, config.name + '_' + contract + ".abi");


        const pathCodeBoc = project.resolve(config.output, config.name + '_' + contract + ".code.boc");
        const pathCodeFif = project.resolve(config.output, config.name + '_' + contract + ".code.fif");
        const pathCodeFifDec = project.resolve(config.output, config.name + '_' + contract + ".code.rev.fif");
        let codeFc: { path: string, content: string }[];
        let codeEntrypoint: string;

        // Compiling contract to func
        logger.log('   > ' + contract + ': tact compiler');
        let abi: string;
        try {
            const res = await compile(ctx, contract, config.name + '_' + contract);
            for (const files of res.output.files) {
                const ffc = project.resolve(config.output, files.name);
                project.writeFile(ffc, files.code);
            }
            project.writeFile(pathAbi, res.output.abi);
            abi = res.output.abi;
            codeFc = res.output.files.map((v) => ({ path: posixNormalize(project.resolve(config.output, v.name)), content: v.code }));
            codeEntrypoint = res.output.entrypoint;
        } catch (e) {
            logger.error('Tact compilation failed');
            logger.error(errorToString(e));
            ok = false;
            continue;
        }

        // Compiling contract to TVM
        logger.log('   > ' + contract + ': func compiler');
        let codeBoc: Buffer;
        try {
            const stdlibPath = stdlib.resolve('stdlib.fc');
            const stdlibCode = stdlib.readFile(stdlibPath).toString();
            const stdlibExPath = stdlib.resolve('stdlib_ex.fc');
            const stdlibExCode = stdlib.readFile(stdlibExPath).toString();
            const c = await funcCompile({
                entries: [
                    stdlibPath,
                    stdlibExPath,
                    posixNormalize(project.resolve(config.output, codeEntrypoint))
                ],
                sources: [{
                    path: stdlibPath,
                    content: stdlibCode
                }, {
                    path: stdlibExPath,
                    content: stdlibExCode,
                },
                ...codeFc
                ],
                logger
            });
            if (!c.ok) {
                logger.error(c.log);
                ok = false;
                continue;
            }
            project.writeFile(pathCodeFif, c.fift);
            project.writeFile(pathCodeBoc, c.output);
            codeBoc = c.output;
        } catch (e) {
            logger.error('FunC compiler crashed');
            logger.error(errorToString(e));
            ok = false;
            continue;
        }

        // Fift decompiler for generated code debug
        logger.log('   > ' + contract + ': fift decompiler');
        let codeFiftDecompiled: string;
        try {
            codeFiftDecompiled = decompileAll({ src: codeBoc });
            project.writeFile(pathCodeFifDec, codeFiftDecompiled);
        } catch (e) {
            logger.error('Fift decompiler crashed');
            logger.error(errorToString(e));
            ok = false;
            continue;
        }

        // Add to built map
        built[contract] = {
            // codeFunc,
            codeBoc,
            // codeFift,
            // codeFiftDecompiled,
            abi
        };
    }
    if (!ok) {
        logger.log('ðŸ’¥ Compilation failed. Skipping packaging');
        return false;
    }

    // Package
    logger.log('   > Packaging');
    const contracts = getContracts(ctx);
    const packages: PackageFileFormat[] = [];
    for (const contract of contracts) {
        logger.log('   > ' + contract);
        const artifacts = built[contract];
        if (!artifacts) {
            logger.error('   > ' + contract + ': no artifacts found');
            return false;
        }

        // System cell
        const depends = Dictionary.empty(Dictionary.Keys.Uint(16), Dictionary.Values.Cell());
        const ct = getType(ctx, contract);
        depends.set(ct.uid, Cell.fromBoc(built[ct.name].codeBoc)[0]); // Mine
        for (const c of ct.dependsOn) {
            const cd = built[c.name];
            if (!cd) {
                logger.error('   > ' + cd + ': no artifacts found');
                return false;
            }
            depends.set(c.uid, Cell.fromBoc(cd.codeBoc)[0]);
        }
        const systemCell = beginCell().storeDict(depends).endCell();

        // Collect sources
        const sources: { [key: string]: string } = {};
        const rawAst = getRawAST(ctx);
        for (const source of [...rawAst.funcSources, ...rawAst.sources]) {
            if (source.path.startsWith(project.root) && !source.path.startsWith(stdlib.root)) {
                sources[source.path.slice(project.root.length)] = Buffer.from(source.code).toString('base64');
            }
        }

        // Package
        const pkg: PackageFileFormat = {
            name: contract,
            abi: artifacts.abi,
            code: artifacts.codeBoc.toString('base64'),
            init: {
                kind: 'direct',
                args: getType(ctx, contract).init!.args.map((v) => ({ name: v.name, type: createABITypeRefFromTypeRef(v.type, v.ref) })),
                prefix: {
                    bits: 1,
                    value: 0,
                },
                deployment: {
                    kind: 'system-cell',
                    system: systemCell.toBoc().toString('base64')
                },
            },
            sources,
            compiler: {
                name: 'tact',
                version: getCompilerVersion(),
                parameters: cfg
            }
        };
        const pkgData = packageCode(pkg);
        const pathPkg = project.resolve(config.output, config.name + '_' + contract + ".pkg");
        project.writeFile(pathPkg, pkgData);
        packages.push(pkg);
    }

    // Bindings
    logger.log('   > Bindings');
    for (const pkg of packages) {
        logger.log('   > ' + pkg.name);
        if (pkg.init.deployment.kind !== 'system-cell') {
            logger.error('   > ' + pkg.name + ': unsupported deployment kind ' + pkg.init.deployment.kind);
            return false;
        }
        try {
            const bindingsServer = writeTypescript(JSON.parse(pkg.abi), {
                code: pkg.code,
                prefix: pkg.init.prefix,
                system: pkg.init.deployment.system,
                args: pkg.init.args
            });
            project.writeFile(project.resolve(config.output, config.name + '_' + pkg.name + ".ts"), bindingsServer);
        } catch (e) {
            logger.error('Bindings compiler crashed');
            logger.error(errorToString(e));
            return false;
        }
    }

    // Reports
    logger.log('   > Reports');
    for (const pkg of packages) {
        logger.log('   > ' + pkg.name);
        try {
            const report = writeReport(ctx, pkg);
            const pathBindings = project.resolve(config.output, config.name + '_' + pkg.name + ".md");
            project.writeFile(pathBindings, report);
        } catch (e) {
            logger.error('Report generation crashed');
            logger.error(errorToString(e));
            return false;
        }
    }

    return true;
}