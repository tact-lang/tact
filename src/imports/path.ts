import { throwInternalCompilerError } from "../error/errors";
import { repeat } from "../utils/array";

// Witness tag. Do not use, do not export.
const pathTag = Symbol("path");

/**
 * Safe relative path
 */
export type RelativePath = {
    /**
     * Number of "../" in front of path
     */
    readonly stepsUp: number;
    /**
     * /-separated strings that go after optional ../
     */
    readonly segments: readonly string[];
    /**
     * Proof that path was created by RelativePath constructor
     */
    readonly [pathTag]: true;
};

/**
 * Constructor for relative paths
 */
const RelativePath = (
    stepsUp: number,
    segments: readonly string[],
): RelativePath => {
    if (stepsUp < 0) {
        throwInternalCompilerError("Negative number of ../ in path");
    }
    const result: RelativePath = {
        [pathTag]: true,
        stepsUp,
        segments: Object.freeze(segments),
    };
    return Object.freeze(result);
};

/**
 * Convert raw string with relative POSIX path into RelativePath
 */
export const fromString = (raw: string): RelativePath => {
    return raw.split("/").map(parseSegment).reduce(appendPath, emptyPath);
};

/**
 * Convert RelativePath to string.
 */
export const asString = ({ stepsUp, segments }: RelativePath): string => {
    return [...repeat("..", stepsUp), ...segments].join("/");
};

/**
 * Empty path, equivalent to "."
 */
export const emptyPath = RelativePath(0, []);

/**
 * Combine two relative paths
 */
const appendPath = (left: RelativePath, right: RelativePath): RelativePath => {
    const delta = right.stepsUp - left.segments.length;
    return RelativePath(left.stepsUp + Math.max(0, delta), [
        ...left.segments.slice(0, Math.max(0, -delta)),
        ...right.segments,
    ]);
};

const parseSegment = (segment: string): RelativePath => {
    if (segment === "..") {
        return RelativePath(1, []);
    }
    if (segment === "." || segment === "") {
        return emptyPath;
    }
    return RelativePath(0, [segment]);
};
