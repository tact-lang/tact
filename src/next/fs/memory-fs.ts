import { appendPath, asString, emptyPath } from "@/next/fs/util";
import { FsErrors } from "@/next/fs/errors";

import type { RelativePath } from "@/next/fs/path";
import type { Cursor } from "@/next/fs/cursor";
import type { Logger } from "@/error/logger-util";

type Options = {
    // FIXME: should we pass log here, or into functions?
    readonly log: Logger<string, void>;

    /**
     * Files present in memory file system at creation time
     */
    readonly files: Map<string, Blob>;

    /**
     * Path to the root, used only for logging
     *
     * Absolute paths do not make sense in memory FS, as
     * it doesn't have the actual root
     */
    readonly root: RelativePath;

    /**
     * Whether it's possible to `write` into
     */
    readonly isReadonly: boolean;
};

export const emptyFiles: Map<string, Blob> = new Map();

/**
 * Create in-memory file system
 */
export function createMemoryFs({
    log,
    files,
    root,
    isReadonly,
}: Options): Cursor {
    const errors = FsErrors(log);

    function builder(currPath: RelativePath): Cursor {
        const getAbsolutePathForLog = () => {
            return asString(appendPath(root, currPath));
        };
        return {
            getAbsolutePathForLog,
            focus: (path) => {
                const newPath = appendPath(currPath, path);
                if (newPath.stepsUp !== 0) {
                    errors.outOfRoot(getAbsolutePathForLog());
                }
                return builder(newPath);
            },
            read: async () => {
                const fullPath = asString(appendPath(root, currPath));
                const blob = files.get(fullPath);
                if (typeof blob === "undefined") {
                    errors.regular.ENOENT(getAbsolutePathForLog());
                    return;
                }
                return new TextDecoder("utf-8").decode(
                    await blob.arrayBuffer(),
                );
            },
            // eslint-disable-next-line @typescript-eslint/require-await
            write: async (content) => {
                if (isReadonly) {
                    errors.readonly(getAbsolutePathForLog());
                }
                const fullPath = asString(appendPath(root, currPath));
                const blob =
                    typeof content === "string"
                        ? new Blob([content], {
                              type: "text/plain;charset=utf-8",
                          })
                        : content;
                files.set(fullPath, blob);
            },
        };
    }

    return builder(emptyPath);
}
