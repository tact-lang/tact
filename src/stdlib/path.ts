import path from "path";
import { posixNormalize } from "../utils/filePath";

export const stdlibPath = posixNormalize(path.join(__dirname, "stdlib"));
