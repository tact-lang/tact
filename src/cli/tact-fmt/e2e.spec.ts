import { join, normalize } from "path";
import { runCommand } from "@/cli/test-util.build";
import { readFileSync, writeFileSync } from "fs";
import { mkdir } from "fs/promises";

// disable tests on windows
const testExceptWindows =
    process.platform === "win32" && process.env.CI ? test.skip : test;

const tactFmt = (...args: string[]) => {
    const unbocPath = normalize(
        join(__dirname, "..", "..", "..", "bin", "tact-fmt.js"),
    );
    const command = `node ${unbocPath} ${args.join(" ")}`;
    return runCommand(command);
};

const outputDir = join(__dirname, "output");

const goodContract = `
contract Test {
    get fun greeting(): String {
        return "hello world";
    }
}
`;

const contractWithSyntaxError = `
contract Test {
    get fun greeting(): String {
        return foo("hello world";
    }
}
`;

describe("tact-fmt foo.tact", () => {
    testExceptWindows("Exits with correct code", async () => {
        await mkdir(outputDir, { recursive: true });
        const file = join(outputDir, "contact.tact");
        writeFileSync(file, goodContract);
        const result = await tactFmt(file);
        expect(result).toMatchObject({ kind: "exited", code: 0 });
    });

    testExceptWindows("Default run", async () => {
        await mkdir(outputDir, { recursive: true });
        const file = join(outputDir, "contact.tact");
        writeFileSync(file, goodContract);
        const result = await tactFmt(file);
        expect(result).toMatchSnapshot();
    });

    testExceptWindows("Default run with write to file", async () => {
        await mkdir(outputDir, { recursive: true });
        const file = join(outputDir, "contact.tact");
        writeFileSync(file, goodContract);
        const result = await tactFmt(file, "-w");
        expect(result).toMatchSnapshot();

        const newContent = readFileSync(file, "utf8");
        expect(newContent).toMatchSnapshot();
    });

    testExceptWindows("With syntax error", async () => {
        await mkdir(outputDir, { recursive: true });
        const file = join(outputDir, "contact.tact");
        writeFileSync(file, contractWithSyntaxError);
        const result = await tactFmt(file, "-w");
        expect(result).toMatchSnapshot();
    });
});
