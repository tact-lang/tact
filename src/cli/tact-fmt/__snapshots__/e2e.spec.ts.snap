// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tact-fmt foo.tact Check and write flags simultaneously 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "",
  "stdout": "Cannot use '--check' and '--write' simultaneously
",
}
`;

exports[`tact-fmt foo.tact Check on directory with formatted files 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
All Tact files use official code style!
",
}
`;

exports[`tact-fmt foo.tact Check on directory with not formatted files 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
[warn] file1.tact
[warn] file2.tact
[warn] file3.tact
Code style issues found in the above file. Run tact-fmt with --write to fix.
",
}
`;

exports[`tact-fmt foo.tact Check on several directories 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
[warn] file2.tact
Code style issues found in the above file. Run tact-fmt with --write to fix.
",
}
`;

exports[`tact-fmt foo.tact Check on several formatted files 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "Checking formatting...
All Tact files use official code style!
",
}
`;

exports[`tact-fmt foo.tact Default run 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": "contract Test {
    get fun greeting(): String {
        return "hello world";
    }
}

",
}
`;

exports[`tact-fmt foo.tact Default run with write to file 1`] = `
"contract Test {
    get fun greeting(): String {
        return "hello world";
    }
}
"
`;

exports[`tact-fmt foo.tact Run on directory 1`] = `
"fun foo1() {}
"
`;

exports[`tact-fmt foo.tact Run on directory 2`] = `
"fun foo2() {}
"
`;

exports[`tact-fmt foo.tact Run on directory 3`] = `
"fun foo3() {}
"
`;

exports[`tact-fmt foo.tact With syntax error 1`] = `
{
  "code": 1,
  "kind": "exited",
  "stderr": "Cannot format file src/cli/tact-fmt/output/contact.tact: src/cli/tact-fmt/output/contact.tact:4:33: Expected "!!", "!=", "%", "&", "&&", "(", ")", "*", "+", ",", "-", ".", "/", "<", "<<", "<=", "==", ">", ">=", ">>", "?", "^", "|", or "||"
  3 |     get fun greeting(): String {
> 4 |         return foo("hello world";
                                      ^
  5 |     }

",
  "stdout": "",
}
`;
